### PR TITLE
Refactor worldgen and add de/serialisation

### PR DESCRIPTION
**Pull request by SuscipiamSingularitatem**

### CODE REVIEW

**Comment by Krutonium** [comment]
For saving and loading would it not be better to use the built in functionality to save the game?

https://docs.godotengine.org/en/stable/tutorials/io/saving_games.html

**Comment by SuscipiamSingularitatem** [comment]
Apart from the use of the GDScript function `String to_json ( Variant var )` that looks like what we want.

File: scripts/GenChunks.cs [hunk 1]
```diff
@@ -51,31 +45,21 @@ public override void _Ready()
...
...⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠑⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠀⠀⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠀⠀⠀⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠀⠀⠀⠀⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠀⠀⠀⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠀⠀⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠓⠀⠀⠓⠓⠓⠓⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠔⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠔⠔⠔⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠔⠔⠔⠔⠔⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀",
```
Comment (Krutonium) [inline scripts/GenChunks.cs:59]: Did you mean to leave this here?

File: scripts/Chunk.cs [hunk 2]
```diff
@@ -1,37 +1,97 @@
 using System;
```
Comment (Krutonium) [inline scripts/Chunk.cs:1]: Unrelated to line:

Could you go through and add comments to the code to explain what's actually going on, for future people who may be working on this code?

File: scripts/GenChunks.cs (hunk 1 above)
Comment (SuscipiamSingularitatem) [reply scripts/GenChunks.cs:59]: It doesn't actually work so this is here for debugging.

File: scripts/Chunk.cs (hunk 2 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:1]: 👌

**Comment by SuscipiamSingularitatem** [comment]
Done~

File: scripts/Chunk.cs [hunk 3]
```diff
@@ -1,37 +1,97 @@
...
+			i / (Chunk.HEIGHT * Chunk.WIDTH));
+
+		private Serialiser() {}
+
+// ReSharper disable MemberHidesStaticFromOuterClass
+		public Chunk Deserialise(string s, (Node2D Parent, char ID) extra)
+// ReSharper restore MemberHidesStaticFromOuterClass
+		{
+			var sSpan = s.AsSpan();
+			OpenSimplexNoise rng = new(); //TODO
```
Comment (Krutonium) [inline scripts/Chunk.cs:31]: Still TODO?

Comment (Krutonium) [reply scripts/Chunk.cs:31]: Also: Commentsssssssss

**Review by Krutonium — COMMENTED** [review]
For the love of god, Please comment your code. What is each part doing? How?

**Comment by Krutonium** [comment]
We could merge this, now, if we knew what it did.

Write. Comments.

Writing comments in a month or a week means there will never be any comments, because you'll forget to, and then when you finally do get around to doing it, you will have exactly as much of an idea of what it does as me and Resin. That just leads to nobody knowing how it works and so much repeated work as we re-write working code to figure out wtf it does.

File: scripts/Chunk.cs [hunk 4]
```diff
@@ -1,37 +1,113 @@
...
+					data[I(x, y, h)] = c.GetTile(x, y, h);
+
+			// reinterpret cast
+			var dataBytes = new byte[data.Length * 4];
+			Buffer.BlockCopy(data, 0, dataBytes, 0, dataBytes.Length);
+
+			var elevStr = ((byte) c.NumElevations).ToString("X2");
+			var dataStr = Convert.ToBase64String(dataBytes);
+			return elevStr + dataStr;
+		}
```
Comment (HazmatDrone) [inline scripts/Chunk.cs:63]: Please explain in more detail.

File: scripts/Chunk.cs [hunk 5]
```diff
@@ -76,15 +154,52 @@ public Chunk(YSort parent, char id, int numElevations)
...
+	}
+
+	/// <returns>tile ID of tile at chunk-relative position and elevation</returns>
+	public readonly int GetTile(int x, int y, int h)
+		=> Terrains[h].GetCell(x, y);
+
+	private readonly int RandomElev(float x, float y)
+	{
+		var f = _rng.GetNoise2d(x, y);
+		return (int) (((f < 0.0f ? -1.0f : 1.0f) * f * f + 1.0f) * NumElevations / 2.0f);
```
Comment (HazmatDrone) [inline scripts/Chunk.cs:184]: What does this do specifically? How can I tweak it if I want to alter the worldgen?

File: scripts/ICustomSerialiser.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/// <typeparam name="TObj">type which is de/serialised</typeparam>
+/// <typeparam name="TSer">extra info used by serialiser (can be tuple for multiple)</typeparam>
+/// <typeparam name="TDSer">extra info used by deserialiser (can be tuple for multiple)</typeparam>
+public interface ICustomSerialiser<TObj, TSer, TDSer>
+	where TObj : notnull
+{
+	TObj Deserialise(string s, TDSer extra);
+
+	string Serialise(TObj c, TSer extra);
+}
```
Comment (HazmatDrone) [inline scripts/ICustomSerialiser.cs:10]: Fantastic, but what does it actually do?

File: scripts/Chunk.cs [hunk 7]
```diff
@@ -56,12 +132,14 @@ public Chunk(YSort parent, char id, int numElevations)
 		});
 	}
 
+	/// <summary>adds {this} chunk's managed <see cref="TileMap">TileMaps</see> to {this} chunk's <see cref="_parent">parent</see></summary>
 	public readonly void Attach()
 	{
 		foreach (var tm in FloorCollisions) _parent.AddChild(tm);
 		foreach (var tm in Terrains) _parent.AddChild(tm);
 	}
```
Comment (HazmatDrone) [inline scripts/Chunk.cs:141]: Please use the word tilemap in full, rather than abbreviations.

File: scripts/GenChunks.cs [hunk 8]
```diff
@@ -51,31 +45,26 @@ public override void _Ready()
...
-					chunk.Set(x, y, h, data[I(tileOffsetX + x, tileOffsetY + y, h)]);
-			chunk.Bump();
-		}
+		for (var cY = 0; cY < 4; cY++) for (var cX = 0; cX < 5; cX++)
+			_chunks[cY * 5 + cX].Run(chunk =>
+			{
+				chunk.SetChunkOffset(cX, cY);
+				chunk.Generate(tileIDs);
+				chunk.Bump();
+			});
```
Comment (HazmatDrone) [inline scripts/GenChunks.cs:54]: Please elaborate on what this does

File: scripts/Chunk.cs [hunk 9]
```diff
@@ -1,37 +1,113 @@
...
+			return c;
+		}
+
+		public string Serialise(Chunk c, ValueTuple extra = default)
+		{
+			// copy tile IDs into int array
+			var data = new int[Chunk.WIDTH * Chunk.HEIGHT * c.NumElevations];
+			for (var h = 0; h < c.NumElevations; h++)
+				for (var y = 0; y < Chunk.HEIGHT; y++) for (var x = 0; x < Chunk.WIDTH; x++)
+					data[I(x, y, h)] = c.GetTile(x, y, h);
```
Comment (HazmatDrone) [inline scripts/Chunk.cs:54]: Turns 3D array into 1D array?

File: scripts/Chunk.cs [hunk 10]
```diff
@@ -1,37 +1,113 @@
...
+
+			// reinterpret cast
+			var dataBytes = new byte[data.Length * 4];
+			Buffer.BlockCopy(data, 0, dataBytes, 0, dataBytes.Length);
```
Comment (HazmatDrone) [inline scripts/Chunk.cs:58]: What does this do?

File: scripts/Chunk.cs [hunk 11]
```diff
@@ -1,37 +1,113 @@
...
+
+			var elevStr = ((byte) c.NumElevations).ToString("X2");
+			var dataStr = Convert.ToBase64String(dataBytes);
+			return elevStr + dataStr;
```
Comment (HazmatDrone) [inline scripts/Chunk.cs:62]: What are these?

**Review by HazmatDrone — COMMENTED** [review]
What does this serialisation scheme give us compared to Godot's builtin that makes it critical to use? 

Please also write out the specs of your serialisation scheme.

File: scripts/ICustomSerialiser.cs (hunk 6 above)
Comment (SuscipiamSingularitatem) [reply scripts/ICustomSerialiser.cs:10]: This is an [interface](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/interfaces). There's a working serialiser [here](https://github.com/HazDroneCorp/combat-engineers/blob/a7265d68339217762251309d71d9dcb7119199ef/scripts/Chunk.cs#L13).

File: scripts/Chunk.cs (hunk 5 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:184]: You'd change the RNG, which is initialised elsewhere and passed to the `Chunk` via its constructor.

This function takes the RNG value (which is in [-1..1]), applies [`g(f) = sign(f) * f^2`](https://www.wolframalpha.com/input/?i=g%28f%29+%3D+sign%28f%29+*+f%5E2+from+f+%3D+-1+to+f+%3D+1), then translates and scales it from [-1..1] to [0..NumElevations). I don't think the squaring part helps much with only 5 elevations.

File: scripts/Chunk.cs (hunk 7 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:141]: `tm` is a loop variable, which conventionally use 1 or 2 character identifiers. I won't be changing this.

File: scripts/GenChunks.cs (hunk 8 above)
Comment (SuscipiamSingularitatem) [reply scripts/GenChunks.cs:54]: Loops through each chunk in `_chunks`, calling the methods `SetChunkOffset`, `Generate`, and `Bump` on each. These set the chunk's offset, generate tiles with the RNG, and update connected textures, respectively, according to the docs on each method.

File: scripts/Chunk.cs (hunk 9 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:54]: Yep

File: scripts/Chunk.cs (hunk 11 above)
Comment (HazmatDrone) [reply scripts/Chunk.cs:62]: Also, what is c?

File: scripts/Chunk.cs (hunk 10 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:58]: A "reinterpret cast" is a C/C++ idiom for manipulating the bytes of an array. In this case, each `int` in `data` becomes 4 `byte`s in `dataBytes`.

File: scripts/Chunk.cs (hunk 5 above)
Comment (HazmatDrone) [reply scripts/Chunk.cs:184]: So, I can change what it returns however I please, so long as it is constrained to an int between 1 and 5?

File: scripts/Chunk.cs (hunk 11 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:62]: These are the two strings which are concatenated to form the final string which is the serialised form of the given `Chunk`. This mirrors the deserialisation method above.

File: scripts/GenChunks.cs (hunk 8 above)
Comment (HazmatDrone) [reply scripts/GenChunks.cs:54]:
```suggestion
		/// <summary> Loops through each chunk in _chunks, setting the chunk's offset, generating tiles with the RNG, and updating connected textures, respectively.</summary>
		for (var cY = 0; cY < 4; cY++) for (var cX = 0; cX < 5; cX++)
			_chunks[cY * 5 + cX].Run(chunk =>
			{
				chunk.SetChunkOffset(cX, cY);
				chunk.Generate(tileIDs);
				chunk.Bump();
			});
```

File: scripts/Chunk.cs (hunk 5 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:184]: 0 and 4. Check the callsite to figure out what returning 5 would do, I imagine it'd throw.

Comment (HazmatDrone) [reply scripts/Chunk.cs:184]: Can you make some exports that would allow me to tweak the worldgen from Godot?

File: scripts/GenChunks.cs (hunk 8 above)
Comment (SuscipiamSingularitatem) [reply scripts/GenChunks.cs:66]: That was supposed to be sarcastic, so to clarify:
```csharp
// loops through each element in list, attempting to un-break each and counting how many remain broken
var broken = 0;
foreach (var e in list)
{
    e.TryFix();
    if (e.IsBroken) broken++;
}
```
This is the kind of comment that I delete, and I refuse to write one.

File: scripts/Chunk.cs (hunk 4 above)
Comment (SuscipiamSingularitatem) [reply scripts/Chunk.cs:63]: Do you really need more explanation than what you have so far?

**Pull request merged**